Change truncation to rounding to estimate frameNumber; Improve backward seeking

anotherche opened the pull request:
So, there are two topics in this request.
The first is quite simple - current code of estimation of frameNumber (made in `public Frame grabFrame(boolean doAudio, boolean doVideo, boolean processImage, boolean keyFrames)`) and `getLengthInFrames()` use simple truncation of the result like 
```
(int)(timestamp * getFrameRate() / 1000000L)
```
or
```
(int)(getLengthInTime() * getFrameRate() / 1000000L)
```
This truncation gives incorrect results if the framerate is not a divisor of 1000000. In this case, the frameNumber is not increasing monotonously and the total number of frames in the video is reduced by one. The problem is solved by changing of the truncation to rounding.

The second topic is about seeking  backwards. The current code use the following while loop in `public void setTimestamp(long timestamp)` function:
```
int count = 0; // prevent infinite loops with corrupted files
while (this.timestamp > timestamp + 1 && grabFrame(true, true, false, false) != null && count++ < 1000) {
                // flush frames if seeking backwards
}
```
If I'm not wrong, this loop looks superfluous in all cases of normal not corrupted files. I think so because the preceding code should guarantee reposition of the decoding place just before the necessary seek and flushing of all internal buffers. Thus, the first successful call of ` grabFrame(true, true, false, false)` should grab a frame with `this.timestamp <= timestamp +1`. It means that, in case of not corrupted files, the while loop will always run exactly once when seeking backwards.  Thus it can be replaced with a single conditional call of the `grabFrame(true, true, false, false)` like 
```
if (this.timestamp > timestamp + 1) grabFrame(true, true, false, false);
```
This change, if it will not break the seeking logic, will additionally improve a situation with some cases of not so good files. I mean that if the previous call of 
```
avformat_seek_file(oc, -1, Long.MIN_VALUE, timestamp, Long.MAX_VALUE, AVSEEK_FLAG_BACKWARD)
```
will not position the decoding to a point **before** the desired timestamp, but position it somewhere **after**, the discussed while loop will bring us at least 1000 frames ahead. If to change the while loop to the conditional grab we will stay closer to the desired seek. I checked the behavior in several tests with movies encoded with various encoders. It seems that the suggested code works well (although I only looked at video frames, not audio samples, but I think there should not be a problem (???)). And it prevent wrong seeking (seeking to 1000 frames ahead) in case of video files encoded with XviD with "packed b-frames" option (which is is not so rear case).

saudet commented:
Thank you! Unfortunately, `FrameGrabberTest.testFFmpegFrameGrabberSeeking()` is failing. Could you check it out and see what is wrong either with your changes or with the test?

anotherche commented:
I'm trying to run the tests, but unfortunately I do not have an experience in junit usage (if it relates to the question at all). 
while trying to run junit test in eclipse I obtain 
```
java.lang.NoClassDefFoundError: Could not initialize class org.bytedeco.javacpp.avutil
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.bytedeco.javacpp.Loader.load(Loader.java:932)
	at org.bytedeco.javacpp.Loader.load(Loader.java:873)
	at org.bytedeco.javacpp.avformat$Write_packet_Pointer_BytePointer_int.<clinit>(avformat.java:640)
	at org.bytedeco.javacv.FFmpegFrameRecorder.<clinit>(FFmpegFrameRecorder.java:309)
	at org.bytedeco.javacv.FrameGrabberTest.testFFmpegFrameGrabberSeeking(FrameGrabberTest.java:272)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:678)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
```
and 

```
java.lang.NoClassDefFoundError: Could not initialize class org.bytedeco.javacv.FFmpegFrameRecorder
	at org.bytedeco.javacv.FrameGrabberTest.testFFmpegFrameGrabber(FrameGrabberTest.java:52)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:678)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)

```

saudet commented:
You'll need to install JavaCV before running tests like that, but other
than that it should work the same as other Maven projects.

saudet commented:
If it doesn't work from Eclipse, just do it on the command line:
```
mvn -U clean install
mvn -U -f platform clean install
```
Eclipse is known to have many issues with Maven...

anotherche commented:
Thank you! It works outside eclipse. 
I see that my idea with changing while loop to a single conditional was wrong in that the while loop cannot run more than once during rewind. 
I replaced "if" to "while" again and checked how it works. Because of reading video and audio frames the first loop runs at least twice to rewind. After that, the second while loop places decoding slightly later the required timestamp (<1000000/framerate).  But sometimes it runs even 30 or 31 times on this generated video. And in those cases the timestamp2 appears almost 1 sec later (>900000). 
If the while loop is replaced with "if" the result is worse. Rewind leads to timestamps located way before the required position, up to 9 seconds. So, on the whole, the replacement of "while" to "if" is not good.

saudet commented:
I'm sure there's a better way to do that, and it must be hidden somewhere in the code of ffplay. If you can take the time to study the way ffplay does seeking, we should be able to understand the best way to do this.

anotherche commented:
So, I improved the seeking code. The tests are going without errors now. I have checked it on various video files, as well as on that generated by the test. The test code is changed somewhat as well. I tried to add comments to every new parts of the code so that it should be self-explanatory more or less.
Additionally, I suggest to add some helper API useful for user applications: information about streams and format.

saudet commented:
Thanks for the changes, but I'm pretty sure this won't work for files with audio streams. I guess I'll be updating the test to make it fail with your changes...

anotherche commented:
Do you mean files containing ONLY audio and no video - I did not checked such (although I tried to account for this possibility). As for files containing audio along with video it works fine. I did detailed tests showing what type of data is containing in every next frame during the seek and what is the timestapms of those intermediate frames.

anotherche commented:
Yes, the previous code was not good for "audio only " files (because of frameRate=0.0). Improved.
New seek code is checked for various types of files: audio+video, video only, audio only (even in ordinary mp3). 
(strange, previous commit for setting frameGrabbed back to false after grabFrame seem to disappear, so that I have to commit it once again)

saudet commented:
Ok, I updated the test, such that it also checks the next audio/video frame after seeking (considering that the first frame is a video/audio frame). Please make sure your changes also work for this case! Thanks

anotherche commented:
@saudet, OK, intriguing.. but I cannot understand the reason why we need do this in the test
```
 while ((wasVideo && frame2.image != null)
                    || (wasAudio && frame2.samples != null)) {
                frame2 = grabber.grab();
            }
```
This loop, as I can see, will 100% make timestamp3 to run away from the initial requested timestamp sooner or later. 
I see this like that: say we grabbed video frame after the seek. Then this loop will continue grab video frames until an audio frame will happen. So if there are many video frames going after in a continuous manner we will run away from the initial seek position. The similar thing is for the case of  wasAudio=true. And another thought is that it will be always so that the `timestamp` and `timestamp3` will always correspond to different types of frames, one is for video and another for audio. I'm not sure if I'm absolutely right, but afaik there's no tight placement of audio and video streams in modern containers like mp4 or mkv. So it can be quite possible that when grabber receives audio frame after a video (or wise-versa) they will have not so close timestamps. Isn't it? Please explain if I'm wrong. 
One thing I will tell for sure. When I played with the test last days I logged sequences of frames (types, timestamps) grabbed during the seek procedure. I saw how series of video frames went after series of audio frames and their timestamps were growing monotonously within one type of frames but jumped when one type of frame changed other. The reason of these jumps is quite  obvious. So this part of test will only succeed with types of containers where audio and video are interleaved  tightly. 
We can check it simply. Let's not seek at all to exclude possible  influence of the seek code. Let's grab repeatedly for a while from the grabber start to a random position (and even better, to a position where one type of frame changes to a different type),  save the timestamp, then grab in the same loop as above, till another frame type change and compare timestamp3 to the previously saved. 

So, it can be like this:
```
grabber.start();
Frame frame2=grabber.grab();
long ts1=grabber.getTimestamp();
boolean wasVideo=frame2.image!=null;
boolean wasAudio=frame2.samples!=null;
int count=0;
long ts2;
while(count++<1000 && frame!=null){
    while ((wasVideo && frame2.image != null)
                    || (wasAudio && frame2.samples != null)) {
                grabber.grab();
            }
   ts2 = grabber.getTimestamp();
    System.out.println(ts2 + " - " + ts1 + " = " + (ts2 - ts1));
            assertTrue(ts2 >= ts1 - 10000000 && ts2 < ts1 + 1000000);
}
```

saudet commented:
The whole point of seeking is to end up in a position where all streams are approximately around the same timestamp. In the case of MP4 files generated by FFmpeg and x264, it looks like the time difference between streams is at most 1 second, by default. Corrupted video files can return anything obviously, but we're not testing for corrupted files here. Most of the time, the index is correct, and for files with both audio and video streams, we should provide some API to provide seeking for both the audio and video streams.

anotherche commented:
So, I've done that test from the previous message. The result is not exactly as I supposed, it is close to that you say about 1 sec. But as well, not exactly less that 1 sec.
The code was:
```
public static final int FRAME_TYPE_NULL = -1;
public static final int FRAME_TYPE_VIDEO = 0;
public static final int FRAME_TYPE_AUDIO = 1;
public static final int FRAME_TYPE_NONE = 2;
	
public int getFrameType(Frame frame) {
	return frame == null? FRAME_TYPE_NULL 
				: frame.image != null? FRAME_TYPE_VIDEO 
				: frame.samples != null? FRAME_TYPE_AUDIO 
				: FRAME_TYPE_NONE;
}
	
public String frameTypeName(int frameType) {
	return frameType == FRAME_TYPE_NULL?"null"
	    		: frameType == FRAME_TYPE_NONE?"none"
	    		: frameType == FRAME_TYPE_VIDEO?"video"
	    		: frameType == FRAME_TYPE_AUDIO?"audio"
	    		: "unknown";
}

. . . 
. . . 
@Test public void testFFmpegFrameGrabberSeeking() throws IOException {
. . . 
. . . 
FFmpegFrameGrabber grabber = new FFmpegFrameGrabber(tempFile);
grabber.start();
long mindelta = Long.MAX_VALUE;
long maxdelta = Long.MIN_VALUE;
Frame frame2=grabber.grab();
long ts1=grabber.getTimestamp();
int frameType = getFrameType(frame2);
int count=0;
long ts2;
while(count++ < 1000  &&  getFrameType(frame2) != FRAME_TYPE_NULL){
	int count1 = 0;
	while (frameType == getFrameType(frame2)) {
		grabber.grab();
		count1++;
	}
	ts2 = grabber.getTimestamp();
	String oldType = frameTypeName(frameType);
	String newType = frameTypeName(getFrameType(frame2));
	long delta = ts2 - ts1;
	if (delta > maxdelta) maxdelta = delta;
	if (delta < mindelta) mindelta = delta;
	System.out.println(ts2 + " - " + ts1 + " = " + delta + " change to "+ newType +" after "+ count1 +" of "+oldType);
	ts1 = ts2;
	frameType = getFrameType(frame2);
}
System.out.println("delta from " + mindelta + " to " + maxdelta);
```

The output based on the video generated in the test was like that
```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.bytedeco.javacv.FrameGrabberTest
FFmpegFrameGrabberSeeking

Seek in file containing video and audio
============================================
Testing file testAV.mp4
Length = 333376000
Framerate = 30.003000300030003

has video stream = YES, has audio stream = YES
0 - 0 = 0 change to audio after 31 of video
1033333 - 0 = 1033333 change to video after 2 of audio
42666 - 1033333 = -990667 change to audio after 1 of video
1066666 - 42666 = 1024000 change to video after 2 of audio
85333 - 1066666 = -981333 change to audio after 1 of video
1100000 - 85333 = 1014667 change to video after 1 of audio
106666 - 1100000 = -993334 change to audio after 1 of video
1133333 - 106666 = 1026667 change to video after 2 of audio
149333 - 1133333 = -984000 change to audio after 1 of video
1166666 - 149333 = 1017333 change to video after 1 of audio
170666 - 1166666 = -996000 change to audio after 1 of video
1200000 - 170666 = 1029334 change to video after 2 of audio
. . . 
. . . 
delta from -1000000 to 1033334
```
That is video goes ~1 sec ahead of audio. And sometimes 1 frame over 1 sec. 
So the test should be changed so that the admissible timestamp difference be something greater than 1 sec. (1.1 sec?)

As for the 
> we should provide some API to perform seeking for both the audio and video streams

The code I suggest seeks both (do you think it does not?). The only feature is that I do so that video frames to have a priority  in seeking over audio. As you say that in normal video files there should not be lags greater than a second between streams, it does not matter for normal usage which type of frame has a priority at seeking. I did this priority to solve my local task ;-) (I needed frame-precision seeking by video frames). So, again, this priority should not have an effect for normal usage (as far as your tests require 1 sec precision only). On the other hand, we could change the API so that there will be a`setTimestamp(long timestamp, int priorityType)` function with a priority selection variable (by video, by audio, regardless of type) and another function having standard signature `setTimestamp(long timestamp)` returning 
 `setTimestamp(timestamp, priorityType=regardless)`

anotherche commented:
By the way, I have made the last test for the full duration of the generated file. The difference between audio and video is ~1 sec almost everywhere in the file. First half the video is ahead and in the second half it is behind the audio. There are only two short periods, in the middle and in the end of the video, where they are very close each other. See the figure

![lags](https://user-images.githubusercontent.com/9344635/36521364-0a6ec362-17c9-11e8-86c5-797d2ee7c49b.png)


So, while

> The whole point of seeking is to end up in a position where all streams are approximately around the same timestamp.

the proximity was actually almost never closer than a second in this indeed not corrupted file.

saudet commented:
Thanks for checking this out, so if your code is supposed to support seeking for both audio and video, the modifications I made to the test should pass. Why does it fail?

saudet commented:
BTW, if you are not interested in audio frames, you should be calling `grabImage()` and not `grab()`.

anotherche commented:
Yes, I use grabImage. But I needed as well a finer tune of the seek. The present code of setTimestamp (I mean the release version, not mine) disregards frame type while seeking to the desired timestamp. So when I need a video frame with a timestamp setTimestamp may give me an audio frame with this timestamp. And then after I do grabImage from this point a video frame can be received with timestamp distanced 1 sec from the desired. That's the problem (this is not a problem for normal video usage, like watching, but this is a problem for picture analysis, when more precise seeking is preferred). That was the main reason I started to change  the setTimestamp function.

>  if your code is supposed to support seeking for both audio and video, the modifications I made to the test should pass. Why does it fail?

That's why:
```
assertTrue(timestamp3 >= timestamp - 10000000 && timestamp3 < timestamp + 1000000);
```
the allowed difference is 1000000, but sometimes little excess is obtained within one frame:
```
Seek in file containing video and audio
============================================
Testing file testAV.mp4
Length = 333376000
Framerate = 30.003000300030003
has video stream = YES, has audio stream = YES
144766666 - 144756186 = 10480
143786666 - 144756186 = -969520
174033333 - 174021057 = 12276
175040000 - 174021057 = 1018943
testFFmpegFrameGrabberSeeking(org.bytedeco.javacv.FrameGrabberTest)  Time elapsed: 286 sec  <<< FAILURE!
```
1018943-1000000=18943 is less than a frame (1000000/30)
in my test above it was obtained `delta from -1000000 to 1033334` between audio and video streams. So the failure check should be changed as
```
assertTrue(timestamp3 >= timestamp - (10000000+tolerance) && timestamp3 < timestamp + (1000000+tolerance));
```
where `tolerance` could be estimated as 2*1000000/frameRate if frameRate is not zero, or simply as 500000 if frameRate is zero

saudet commented:
Ah, I see, you have a good point. So what about adding new `setVideoTimestamp()` and `setAudioTimestamp()` methods? Or actually `videoTimestamp` and `audioTimestamp` properties defined in the `FrameGrabber` parent class that keep track of the timestamps of the last video and audio frames returned?

anotherche commented:
I repaired th seeking test. 
I think I will do the seeking by audio and video soon

anotherche commented:
@saudet, it is more or less ready (there can be some mistakes in comments still). I added all functions to have a possibility to seek (to a timestamp or to a framenumber) by video, by audio, or by any (as in the original API). The only thing I do not like is the way I obtain the fremarate for audio frames which is needed for the seeking by audio frames. The problem is that I do not see a possibility to obtain the number of samples in an audio frame other than `samples_frame.nb_samples()`. But for that to work there should be some samples already grabbed before (otherwise `samples_frame` is null or `samples_frame.nb_samples()` is zero). If there was no samples grabbed I do an unprocessed `grabFrame(true, false, false, false)`. Although I set the `frameGrabbed` flag, anyway it is an additional unnecessary grab (a grab which was not asked by user and which may potentially lead to skipping of some video frames before the audio frame will be reached). I tried in the code that this problem not to appear at all (the audio frame rate is only necessary in cases where there is only  audio stream, so there should be always audio frames grabbed before). But anyway, I would be more happy if there would be a possibility to determine the number of samples per audio frame without grabbing an audio frame. Do you know of such a way?

saudet commented:
You've even added a `setAudioFrameNumber()`, this is great! I'm not sure how useful it is, as you say, but sure why not. We'll let others try it out and see if they can find a better way. However `setVideoFrameNumber()` is missing?

anotherche commented:
Well, it is because the original `setFrameNumber()` was implying that the frame is the video frame, because the timestamp to seek to was calculated as `1000000*frameNumber/getFrameRate()` (getFrameRate() returns the rate of video frames). Thus, actually `setFrameNumber()` was always equivalent to something like `setVideoFrameNumber()`.
Anyway, I agree that it is worth adding the signature of `setVideoFrameNumber()` just for the symmetry. But inside it will do the same as `setFrameNumber()` which must be untouched because it overrides of the super class's method.
I've done that simple addition (and a couple of corrections as well) but I have a strange dependency trouble with compiling  tests in `platform` folder
```
Failed to execute goal on project javacv-platform: Could not resolve dependencies for project
org.bytedeco:javacv-platform:jar:1.4.1-SNAPSHOT: The following artifacts could not be resolved:
org.bytedeco.javacpp-presets:opencv:jar:android-arm64:3.4.0-1.4.1-SNAPSHOT, 
org.bytedeco.javacpp-presets:opencv:jar:android-x86_64:3.4.0-1.4.1-SNAPSHOT: 
Could not find artifact org.bytedeco.javacpp-presets:opencv:jar:android-arm64:3.4.0-1.4.1-SNAPSHOT 
in sonatype-nexus-snapshots
```
how to overcome?

saudet commented:
I've updated the versions of OpenCV and FFmpeg, so it works again now! Sorry about that.

anotherche commented:
Yes, it works again. But I had to fetch from bytedeco javacv first to make it work again. IDK if it was correct (all that maven is still quite new for me), but it seem to work. As I told, I added `setVideoTimestamp()` for the symmetry with `setAudioTimestamp()`. As this `setVideoTimestamp()` should do the same as the original `setTimestamp()` I've made it a kind of alias. As well I removed some occasional and unnecessary code and comments remaining from the intermediate stages :-).
So, basically, I do not see any unresolved problem in this PR for now.

anotherche commented:
Oh! I'm sorry to give the name `setVideoTimestamp()` to the commit [fea0388](https://github.com/bytedeco/javacv/pull/908/commits/fea038806615ae01cdd80b1e056d63026978c71d) It must be called `setVideoFrameNumber()` of course.

saudet commented:
That's fine, thanks! It looks like it's working well now. :+1: Just a couple of more things. Could we add a `getLengthInAudioFrames()` method? And also, let's either make `FrameTypeToSeek` private, or better yet, leave it public, but move it to `Frame.Type` and use `EnumSet` instead of the `mask` and `ANY` hacks. Thanks!

anotherche commented:
> Could we add a getLengthInVideoFrames() method?

Do you mean something similar I did to  `setVideoFrameNumber()`? I ask because the situation is similar: current `getLengthInFrames()` returns length in video frames in fact. As we do not want to break existing user's code which already uses `getLengthInFrames()` we should leave it as is. Then  `setVideoFrameNumber()` will be an alias of `getLengthInFrames()`. I'll do it if you think it makes sense.

> ... but move it to Frame.Type and use EnumSet instead of the mask and ANY hacks.

OK. Thank you for help me to stop using java as C 👍

saudet commented:
Yes, same as setFrameNumber(), and actually, we should do the same with
setTimestamp(), as it used to seek to video frames only too. Please make
that change as well! Thanks

anotherche commented:
Done. I have added Frame.Type enum as you asked to be used as EnumSet in seeking functions. Additionally, I added Frame.getType() method to return EnumSet of Frame.Type enum constants. IDK if a frame may ever contain both audio and video data. The Frame class definition does not forbid a frame to have both image and samples members not null. So I've made the code to account now for this possibility. If you think that it is impossible (if image!=null and samples!=null are absolutely mutually exclusive) then we may change the return type of Frame.getType() from EnumSet to a single  Frame.Type  constant (and then change the code of the seeking functions accordingly).

>  and actually, we should do the same with setTimestamp(), as it used to seek to video frames only too

No, it is actually different. `setTimestamp()` function is seeking to any (video or audio)  frame having the timestamp closest to the requested. So, currently we have three different functions: `setTimestamp()` is seeking to any frame containing video or audio; `setVideoTimestamp()` is seeking to video frames (if there is no video stream it simply works like `setTimestamp()`); `setAudioTimestamp()` is seeking to audio frames (if there is no audio stream it again simply works like `setTimestamp()`);

saudet commented:
Looks good, thanks! Sorry for the delay.
